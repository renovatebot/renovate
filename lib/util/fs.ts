import * as fs from 'fs-extra';
import { join, parse } from 'upath';
import { RenovateConfig } from '../config/common';
import { logger } from '../logger';

let localDir = '';

export function setFsConfig(config: Partial<RenovateConfig>): void {
  localDir = config.localDir;
}

export function getSubDirectory(fileName: string): string {
  return parse(fileName).dir;
}

export function getSiblingFileName(
  existingFileNameWithPath: string,
  otherFileName: string
): string {
  const subDirectory = getSubDirectory(existingFileNameWithPath);
  return join(subDirectory, otherFileName);
}

export async function readLocalFile(fileName: string): Promise<Buffer>;
export async function readLocalFile(
  fileName: string,
  encoding: 'utf8'
): Promise<string>;
export async function readLocalFile(
  fileName: string,
  encoding?: string
): Promise<string | Buffer> {
  const localFileName = join(localDir, fileName);
  try {
    const fileContent = await fs.readFile(localFileName, encoding);
    return fileContent;
  } catch (err) {
    logger.trace({ err }, 'Error reading local file');
    return null;
  }
}

export async function writeLocalFile(
  fileName: string,
  fileContent: string
): Promise<void> {
  const localFileName = join(localDir, fileName);
  await fs.outputFile(localFileName, fileContent);
}

export async function deleteLocalFile(fileName: string): Promise<void> {
  await fs.remove(fileName);
}

// istanbul ignore next
export async function ensureDir(dirName): Promise<void> {
  await fs.ensureDir(dirName);
}

// istanbul ignore next
export async function ensureLocalDir(dirName): Promise<void> {
  const localDirName = join(localDir, dirName);
  await fs.ensureDir(localDirName);
}
