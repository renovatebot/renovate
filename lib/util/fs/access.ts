import upath from 'upath';
import { FILE_ACCESS_VIOLATION_ERROR } from '../../constants/error-messages';
import { logger } from '../../logger';

export function assertBaseDir(path: string, baseDir: string): void {
  if (!path.startsWith(upath.resolve(baseDir))) {
    logger.warn(
      { path, baseDir },
      'Preventing access to file outside the base directory'
    );
    throw new Error(FILE_ACCESS_VIOLATION_ERROR);
  }
}
