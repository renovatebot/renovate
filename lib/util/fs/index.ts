import * as fs from 'fs-extra';
import { join, parse } from 'upath';
import { RenovateConfig } from '../../config/common';
import { logger } from '../../logger';

export * from './proxies';

let localDir = '';
let cacheDir = '';

export function setFsConfig(config: Partial<RenovateConfig>): void {
  localDir = config.localDir;
  cacheDir = config.cacheDir;
}

export function getSubDirectory(fileName: string): string {
  return parse(fileName).dir;
}

export function getSiblingFileName(
  existingFileNameWithPath: string,
  otherFileName: string
): string {
  const subDirectory = getSubDirectory(existingFileNameWithPath);
  return join(subDirectory, otherFileName);
}

export async function readLocalFile(fileName: string): Promise<Buffer>;
export async function readLocalFile(
  fileName: string,
  encoding: 'utf8'
): Promise<string>;
export async function readLocalFile(
  fileName: string,
  encoding?: string
): Promise<string | Buffer> {
  const localFileName = join(localDir, fileName);
  try {
    const fileContent = await fs.readFile(localFileName, encoding);
    return fileContent;
  } catch (err) {
    logger.trace({ err }, 'Error reading local file');
    return null;
  }
}

export async function writeLocalFile(
  fileName: string,
  fileContent: string
): Promise<void> {
  const localFileName = join(localDir, fileName);
  await fs.outputFile(localFileName, fileContent);
}

export async function deleteLocalFile(fileName: string): Promise<void> {
  if (localDir) {
    const localFileName = join(localDir, fileName);
    await fs.remove(localFileName);
  }
}

// istanbul ignore next
export async function ensureDir(dirName: string): Promise<void> {
  await fs.ensureDir(dirName);
}

// istanbul ignore next
export async function ensureLocalDir(dirName: string): Promise<void> {
  const localDirName = join(localDir, dirName);
  await fs.ensureDir(localDirName);
}

export async function ensureCacheDir(
  dirName: string,
  envPathVar?: string
): Promise<string> {
  const envCacheDirName = envPathVar ? process.env[envPathVar] : null;
  const cacheDirName = envCacheDirName || join(cacheDir, dirName);
  await fs.ensureDir(cacheDirName);
  return cacheDirName;
}

/**
 * Return the path of the private cache directory. This directory is wiped
 * between repositories, so they can be used to store private registries' index
 * without risk of that information leaking to other repositories/users.
 */
export function privateCacheDir(): string {
  return join(cacheDir, '__renovate-private-cache');
}

export function localPathExists(pathName: string): Promise<boolean> {
  // Works for both files as well as directories
  return fs
    .stat(join(localDir, pathName))
    .then((s) => !!s)
    .catch(() => false);
}
