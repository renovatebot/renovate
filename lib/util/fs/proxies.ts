import * as fs from 'fs-extra';
import { MoveOptions, WriteFileOptions } from 'fs-extra';

// istanbul ignore next
export function stat(path: string | Buffer): Promise<fs.Stats> {
  return fs.stat(path);
}

// istanbul ignore next
export function chmod(
  path: string | Buffer,
  mode: string | number
): Promise<void> {
  return fs.chmod(path, mode);
}

export async function readFile(fileName: string): Promise<Buffer>;
export async function readFile(
  fileName: string,
  encoding: 'utf8'
): Promise<string>;
export function readFile(
  fileName: string,
  encoding?: string
): Promise<string | Buffer> {
  return fs.readFile(fileName, encoding);
}

// istanbul ignore next
export function writeFile(
  fileName: string,
  fileContent: string
): Promise<void> {
  return fs.writeFile(fileName, fileContent);
}

// istanbul ignore next
export function outputFile(
  file: string,
  data: any,
  options?: WriteFileOptions | string
): Promise<void> {
  return fs.outputFile(file, data, options ?? {});
}

export function remove(dir: string): Promise<void> {
  return fs.remove(dir);
}

// istanbul ignore next
export function unlink(path: string | Buffer): Promise<void> {
  return fs.unlink(path);
}

// istanbul ignore next
export function exists(path: string): Promise<boolean> {
  return fs.pathExists(path);
}

// istanbul ignore next
export function pathExists(path: string): Promise<boolean> {
  return fs.pathExists(path);
}

// istanbul ignore next
export function move(
  src: string,
  dest: string,
  options?: MoveOptions
): Promise<void> {
  return fs.move(src, dest, options ?? {});
}
