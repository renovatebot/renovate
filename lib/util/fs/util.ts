import upath from 'upath';
import { GlobalConfig } from '../../config/global';
import { FILE_ACCESS_VIOLATION_ERROR } from '../../constants/error-messages';
import { logger } from '../../logger';

function assertBaseDir(path: string, baseDir: string): void {
  if (!path.startsWith(upath.resolve(baseDir))) {
    logger.debug(
      { path, baseDir },
      'Preventing access to file outside the base directory'
    );
    throw new Error(FILE_ACCESS_VIOLATION_ERROR);
  }
}

function ensurePath(path: string, key: 'localDir' | 'cacheDir'): string {
  const baseDir = upath.resolve(GlobalConfig.get(key)!);
  const fullPath = upath.resolve(
    upath.isAbsolute(path) ? path : upath.join(baseDir, path)
  );
  assertBaseDir(fullPath, baseDir);
  return fullPath;
}

export function ensureLocalPath(path: string): string {
  return ensurePath(path, 'localDir');
}

export function ensureCachePath(path: string): string {
  return ensurePath(path, 'cacheDir');
}
