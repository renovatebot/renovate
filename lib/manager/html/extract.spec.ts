import { readFileSync } from 'fs';
import { resolve } from 'upath';
import { extractPackageFile } from '.';

const sample = readFileSync(
  resolve(__dirname, `./__fixtures__/sample.html`),
  'utf8'
);
const nothing = readFileSync(
  resolve(__dirname, `./__fixtures__/nothing.html`),
  'utf8'
);

describe('manager/html/extract', () => {
  it('extractPackageFile', () => {
    expect(extractPackageFile(sample)).toMatchSnapshot();
  });
  it('returns null', () => {
    expect(extractPackageFile(nothing)).toBeNull();
  });
});
