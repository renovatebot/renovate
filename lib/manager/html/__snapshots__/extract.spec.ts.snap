// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/html/extract extractPackageFile 1`] = `
Object {
  "deps": Array [
    Object {
      "currentValue": "15.6.1",
      "datasource": "cdnjs",
      "depName": "prop-types",
      "lookupName": "prop-types/prop-types.min.js",
      "replaceString": "<script type=\\"text/javascript\\"
            src=\\"https://cdnjs.cloudflare.com/ajax/libs/prop-types/15.6.1/prop-types.min.js\\">",
    },
    Object {
      "currentValue": "16.3.2",
      "datasource": "cdnjs",
      "depName": "react",
      "lookupName": "react/umd/react.production.min.js",
      "replaceString": "<script type=\\"text/javascript\\"
            src=\\"https://cdnjs.cloudflare.com/ajax/libs/react/16.3.2/umd/react.production.min.js\\">",
    },
    Object {
      "currentValue": "16.3.2",
      "datasource": "cdnjs",
      "depName": "react-dom",
      "lookupName": "react-dom/umd/react-dom.production.min.js",
      "replaceString": "<script type=\\"text/javascript\\"
            src=\\"https://cdnjs.cloudflare.com/ajax/libs/react-dom/16.3.2/umd/react-dom.production.min.js\\">",
    },
    Object {
      "currentValue": "2.2.1",
      "datasource": "cdnjs",
      "depName": "react-transition-group",
      "lookupName": "react-transition-group/react-transition-group.min.js",
      "replaceString": "<script type=\\"text/javascript\\"
            src=\\"https://cdnjs.cloudflare.com/ajax/libs/react-transition-group/2.2.1/react-transition-group.min.js\\">",
    },
    Object {
      "currentValue": "1.14.3",
      "datasource": "cdnjs",
      "depName": "popper.js",
      "lookupName": "popper.js/umd/popper.min.js",
      "replaceString": "<script type=\\"text/javascript\\"
            src=\\"https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js\\">",
    },
    Object {
      "currentValue": "0.10.4",
      "datasource": "cdnjs",
      "depName": "react-popper",
      "lookupName": "react-popper/umd/react-popper.min.js",
      "replaceString": "<script type=\\"text/javascript\\"
            src=\\"https://cdnjs.cloudflare.com/ajax/libs/react-popper/0.10.4/umd/react-popper.min.js\\">",
    },
    Object {
      "currentValue": "7.1.0",
      "datasource": "cdnjs",
      "depName": "reactstrap",
      "lookupName": "reactstrap/reactstrap.min.js",
      "replaceString": "<script src=\\"https://cdnjs.cloudflare.com/ajax/libs/reactstrap/7.1.0/reactstrap.min.js\\">",
    },
    Object {
      "currentValue": "4.3.1",
      "datasource": "cdnjs",
      "depName": "react-router",
      "lookupName": "react-router/react-router.min.js",
      "replaceString": "<script src=\\" https://cdnjs.cloudflare.com/ajax/libs/react-router/4.3.1/react-router.min.js\\">",
    },
    Object {
      "currentValue": "4.0.6",
      "datasource": "cdnjs",
      "depName": "react-markdown",
      "lookupName": "react-markdown/react-markdown.js",
      "replaceString": "<script src=\\"https://cdnjs.cloudflare.com/ajax/libs/react-markdown/4.0.6/react-markdown.js\\">",
    },
    Object {
      "currentDigest": "sha256-mpnrJ5DpEZZkwkE1ZgkEQQJW/46CSEh/STrZKOB/qoM=",
      "currentValue": "0.18.0",
      "datasource": "cdnjs",
      "depName": "axios",
      "lookupName": "axios/axios.min.js",
      "replaceString": "<script src=\\"https://cdnjs.cloudflare.com/ajax/libs/axios/0.18.0/axios.min.js\\"
            integrity=\\"sha256-mpnrJ5DpEZZkwkE1ZgkEQQJW/46CSEh/STrZKOB/qoM=\\" crossorigin=\\"anonymous\\">",
    },
  ],
}
`;
